main.py

```diff
@@ -137,8 +142,10 @@ def main():
         logging.error("Error retrieving configuration, exiting")
         sys.exit(1)
 
-    modbus_client = setup_modbus_client(configuration)
-    mqtt_client = setup_mqtt_client(configuration)
+    error_handler = setup_error_handler(configuration)
+    modbus_client = setup_modbus_client(configuration, error_handler)
+    mqtt_client = setup_mqtt_client(configuration, error_handler)
+    error_handler.client = mqtt_client
```

Comment:
you could split the responsibilities though. They don't have to depend on the same thing. Have separate abstractions to subscribe and publish and each having it's own instantiation process.
What do you need from the `mqtt_client`? I think you only care about the connection. We could, hypothetically, be using different connections as well. To subscribe from mosquitto and to write to something else.
I wonder if you can instead just keep them separated from each other. They both would depend on the same library (paho?), but besides that, they share little in common

Reply:
hmm so perhaps instead of a single `MQTTClient` module, we would have an `MQTTReader` and `MQTTWriter` ? seems like it might be a bit overkill - the writer would need to manage its own connection rather than just using the existing one. If we actually want to enable using two different MQTT brokers for commands and errors, then we'd need a different set of MQTT config settings as well. I think unless there's a compelling reason to allow two separate connections, I'd rather keep it simple and stick with the slightly clunky method of passing mqtt client to error handler after initialising.

Reply:
I believe I am a bit biased here by what telegraf does, but I think is worth discussing this further.
In general, I feel that objects should have _all_ their dependencies on instantiation. If you can't do that, is often, if not always, a problem with the design. Now, on saying that, I am not saying that the only way to have a better design is by doing what I am suggesting, but I'd like to avoid the partial state on the constructor as a goal. The how we achieve that, remains open, but I will try to explain my perspective.

<img width="260" alt="image" src="https://github.com/EvergenEnergy/remote-commands-handler/assets/1817347/bbb9ad0e-0a26-49c1-a141-b7b710c6d6c7">

The current class is responsible to act, not as a generic MQTT client, but as a very opinionated, dedicated abstraction to take JSON documents. It has a main `run` method already.

We could make the class objective more generic (renaming the `run` method), but I would argue that would just be reducing cohesion of the class. The class is now doing more disparate functions. I tend to feel that SRP is a nice principle here. On saying that, we could also decide to rename the class/file, to make it more explicit that the objective is to act as the `MQTT ingestor`  instead of a generic abstraction around MQTT.


Telegraf has 4 main abstraction and everything is essentially an implementation of one of those abstraction. You have input data, output data, transformers and serialisers. The inputs and outputs remain separated from each other. I like that approach, as allows them to evolve independently from each other. 

For the longer objectives we have, at some point, we will have to think about multiple device protocol support, not only Modbus. Sure, I am here talking about something else, this abstraction is about MQTT, not Modbus. But probably is a means to explain that we may need to think about having possible more than way to back-propagate error messages, not only via MQTT. I would like to have an abstraction that's responsible to handle errors that's not necessarily coupled to way we receive input data.

Reply:
Also, worth mentioning that is probably incorrect to expose a public method to allow the client to determine where to subscribe. To be clear here, I am talking about a parallel task, not the one you are actively working on, but I would recommend to make private the `subscribe_topics` method
Upon instantiation, once you receive the config file, you can subscribe this abstraction to the correct topic.
No need to let the client orchestrate that.

Reply:
OK, still some tidying up to do, but I now have 3 separate MQTT classes:
- base class `MQTTClient` which only connects
- `MQTTReader` which subscribes to topics and runs a loop
- `MQTTWriter` which connects when needed to publish an error message

It works ok and there's not too much duplication of code, and it means the error handler and mqtt reader can be initialised without circular dependencies in `main.py`. Note this is only because the writer module doesn't take an error handler, on the assumption that if it encounters any errors, we're by definition not going to be able to propagate them via MQTT. If we did want to still pass in an error handler and use other non-mqtt behaviour to publish errors, there would still be a circular dependency problem.

There are some awkward bits, like the fact that to mock the internal MQTT client inside error handler, we have to pass the mock down through two layers (first to error_handler init, then to MQTTWriter init). I probably could refine this to make the 'error reporting mechanism' a bit more abstract, but I'm not sure how much benefit there is in further abstraction. Also there's still some coverage to add and docs to update. Let me know what you think :)